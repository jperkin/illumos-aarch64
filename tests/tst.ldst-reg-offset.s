/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.10 Load/store register (register offset)
 *
 * C5.6.145 PRFM is part of this group but is tested in tst.ldst-prfm.s
 */
.macro ldst_reg_offset_simd op, reg, size
	\op	\reg\()0,  [x0,  x0]
	\op	\reg\()31, [sp,  wzr, uxtw]
	\op	\reg\()0,  [x30, w30, uxtw #\size]
	\op	\reg\()31, [sp,  xzr, lsl  #0]
	\op	\reg\()0,  [x30, x30, lsl  #\size]
	\op	\reg\()31, [sp,  wzr, sxtw]
	\op	\reg\()0,  [x30, w30, sxtw #\size]
	\op	\reg\()31, [sp,  xzr, sxtx]
	\op	\reg\()0,  [x30, x30, sxtx #\size]
.endm

.macro ldst_reg_offset op, wreg=0, xreg=0, simd=0, size32=0, size64=0
	.if \wreg
	\op	w0,  [x0,  x0]
	\op	w30, [x30, x30, sxtx]
	\op	wzr, [sp,  xzr, sxtx #\size32]
	\op	wzr, [sp,  xzr, lsl  #0]
	.endif
	.if \xreg
	\op	x0,  [x0,  x0]
	\op	x30, [x30, w30, uxtw]
	\op	xzr, [sp,  wzr, uxtw #\size64]
	\op	xzr, [sp,  xzr, lsl  #\size64]
	\op	xzr, [sp,  wzr, sxtw #\size64]
	\op	xzr, [sp,  xzr, sxtx #\size64]
	.endif
	.if \simd
	ldst_reg_offset_simd \op b 0
	ldst_reg_offset_simd \op h 1
	ldst_reg_offset_simd \op s 2
	ldst_reg_offset_simd \op d 3
	ldst_reg_offset_simd \op q 4
	.endif
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldst_reg_offset strb,  1, 0, 0, 0, 0
	ldst_reg_offset ldrb,  1, 0, 0, 0, 0
	ldst_reg_offset ldrsb, 1, 1, 0, 0, 0
	ldst_reg_offset str,   1, 1, 1, 2, 3
	ldst_reg_offset ldr,   1, 1, 1, 2, 3
	ldst_reg_offset strh,  1, 0, 0, 1, 1
	ldst_reg_offset ldrh,  1, 0, 0, 1, 1
	ldst_reg_offset ldrsh, 1, 1, 0, 1, 1
	ldst_reg_offset ldrsw, 0, 1, 0, 2, 2
.size libdis_test, [.-libdis_test]
