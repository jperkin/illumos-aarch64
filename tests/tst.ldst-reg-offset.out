    libdis_test:       00 68 20 38  strb w0, [x0, x0]
    libdis_test+0x4:   de eb 3e 38  strb w30, [x30, x30, sxtx]
    libdis_test+0x8:   ff fb 3f 38  strb wzr, [sp, xzr, sxtx #0]
    libdis_test+0xc:   ff 7b 3f 38  strb wzr, [sp, xzr, lsl #0]
    libdis_test+0x10:  00 68 60 38  ldrb w0, [x0, x0]
    libdis_test+0x14:  de eb 7e 38  ldrb w30, [x30, x30, sxtx]
    libdis_test+0x18:  ff fb 7f 38  ldrb wzr, [sp, xzr, sxtx #0]
    libdis_test+0x1c:  ff 7b 7f 38  ldrb wzr, [sp, xzr, lsl #0]
    libdis_test+0x20:  00 68 e0 38  ldrsb w0, [x0, x0]
    libdis_test+0x24:  de eb fe 38  ldrsb w30, [x30, x30, sxtx]
    libdis_test+0x28:  ff fb ff 38  ldrsb wzr, [sp, xzr, sxtx #0]
    libdis_test+0x2c:  ff 7b ff 38  ldrsb wzr, [sp, xzr, lsl #0]
    libdis_test+0x30:  00 68 a0 38  ldrsb x0, [x0, x0]
    libdis_test+0x34:  de 4b be 38  ldrsb x30, [x30, w30, uxtw]
    libdis_test+0x38:  ff 5b bf 38  ldrsb xzr, [sp, wzr, uxtw #0]
    libdis_test+0x3c:  ff 7b bf 38  ldrsb xzr, [sp, xzr, lsl #0]
    libdis_test+0x40:  ff db bf 38  ldrsb xzr, [sp, wzr, sxtw #0]
    libdis_test+0x44:  ff fb bf 38  ldrsb xzr, [sp, xzr, sxtx #0]
    libdis_test+0x48:  00 68 20 b8  str w0, [x0, x0]
    libdis_test+0x4c:  de eb 3e b8  str w30, [x30, x30, sxtx]
    libdis_test+0x50:  ff fb 3f b8  str wzr, [sp, xzr, sxtx #2]
    libdis_test+0x54:  ff 6b 3f b8  str wzr, [sp, xzr]
    libdis_test+0x58:  00 68 20 f8  str x0, [x0, x0]
    libdis_test+0x5c:  de 4b 3e f8  str x30, [x30, w30, uxtw]
    libdis_test+0x60:  ff 5b 3f f8  str xzr, [sp, wzr, uxtw #3]
    libdis_test+0x64:  ff 7b 3f f8  str xzr, [sp, xzr, lsl #3]
    libdis_test+0x68:  ff db 3f f8  str xzr, [sp, wzr, sxtw #3]
    libdis_test+0x6c:  ff fb 3f f8  str xzr, [sp, xzr, sxtx #3]
    libdis_test+0x70:  00 68 20 3c  str b0, [x0, x0]
    libdis_test+0x74:  ff 4b 3f 3c  str b31, [sp, wzr, uxtw]
    libdis_test+0x78:  c0 5b 3e 3c  str b0, [x30, w30, uxtw #0]
    libdis_test+0x7c:  ff 7b 3f 3c  str b31, [sp, xzr, lsl #0]
    libdis_test+0x80:  c0 7b 3e 3c  str b0, [x30, x30, lsl #0]
    libdis_test+0x84:  ff cb 3f 3c  str b31, [sp, wzr, sxtw]
    libdis_test+0x88:  c0 db 3e 3c  str b0, [x30, w30, sxtw #0]
    libdis_test+0x8c:  ff eb 3f 3c  str b31, [sp, xzr, sxtx]
    libdis_test+0x90:  c0 fb 3e 3c  str b0, [x30, x30, sxtx #0]
    libdis_test+0x94:  00 68 20 7c  str h0, [x0, x0]
    libdis_test+0x98:  ff 4b 3f 7c  str h31, [sp, wzr, uxtw]
    libdis_test+0x9c:  c0 5b 3e 7c  str h0, [x30, w30, uxtw #1]
    libdis_test+0xa0:  ff 6b 3f 7c  str h31, [sp, xzr]
    libdis_test+0xa4:  c0 7b 3e 7c  str h0, [x30, x30, lsl #1]
    libdis_test+0xa8:  ff cb 3f 7c  str h31, [sp, wzr, sxtw]
    libdis_test+0xac:  c0 db 3e 7c  str h0, [x30, w30, sxtw #1]
    libdis_test+0xb0:  ff eb 3f 7c  str h31, [sp, xzr, sxtx]
    libdis_test+0xb4:  c0 fb 3e 7c  str h0, [x30, x30, sxtx #1]
    libdis_test+0xb8:  00 68 20 bc  str s0, [x0, x0]
    libdis_test+0xbc:  ff 4b 3f bc  str s31, [sp, wzr, uxtw]
    libdis_test+0xc0:  c0 5b 3e bc  str s0, [x30, w30, uxtw #2]
    libdis_test+0xc4:  ff 6b 3f bc  str s31, [sp, xzr]
    libdis_test+0xc8:  c0 7b 3e bc  str s0, [x30, x30, lsl #2]
    libdis_test+0xcc:  ff cb 3f bc  str s31, [sp, wzr, sxtw]
    libdis_test+0xd0:  c0 db 3e bc  str s0, [x30, w30, sxtw #2]
    libdis_test+0xd4:  ff eb 3f bc  str s31, [sp, xzr, sxtx]
    libdis_test+0xd8:  c0 fb 3e bc  str s0, [x30, x30, sxtx #2]
    libdis_test+0xdc:  00 68 20 fc  str d0, [x0, x0]
    libdis_test+0xe0:  ff 4b 3f fc  str d31, [sp, wzr, uxtw]
    libdis_test+0xe4:  c0 5b 3e fc  str d0, [x30, w30, uxtw #3]
    libdis_test+0xe8:  ff 6b 3f fc  str d31, [sp, xzr]
    libdis_test+0xec:  c0 7b 3e fc  str d0, [x30, x30, lsl #3]
    libdis_test+0xf0:  ff cb 3f fc  str d31, [sp, wzr, sxtw]
    libdis_test+0xf4:  c0 db 3e fc  str d0, [x30, w30, sxtw #3]
    libdis_test+0xf8:  ff eb 3f fc  str d31, [sp, xzr, sxtx]
    libdis_test+0xfc:  c0 fb 3e fc  str d0, [x30, x30, sxtx #3]
    libdis_test+0x100: 00 68 a0 3c  str q0, [x0, x0]
    libdis_test+0x104: ff 4b bf 3c  str q31, [sp, wzr, uxtw]
    libdis_test+0x108: c0 5b be 3c  str q0, [x30, w30, uxtw #4]
    libdis_test+0x10c: ff 6b bf 3c  str q31, [sp, xzr]
    libdis_test+0x110: c0 7b be 3c  str q0, [x30, x30, lsl #4]
    libdis_test+0x114: ff cb bf 3c  str q31, [sp, wzr, sxtw]
    libdis_test+0x118: c0 db be 3c  str q0, [x30, w30, sxtw #4]
    libdis_test+0x11c: ff eb bf 3c  str q31, [sp, xzr, sxtx]
    libdis_test+0x120: c0 fb be 3c  str q0, [x30, x30, sxtx #4]
    libdis_test+0x124: 00 68 60 b8  ldr w0, [x0, x0]
    libdis_test+0x128: de eb 7e b8  ldr w30, [x30, x30, sxtx]
    libdis_test+0x12c: ff fb 7f b8  ldr wzr, [sp, xzr, sxtx #2]
    libdis_test+0x130: ff 6b 7f b8  ldr wzr, [sp, xzr]
    libdis_test+0x134: 00 68 60 f8  ldr x0, [x0, x0]
    libdis_test+0x138: de 4b 7e f8  ldr x30, [x30, w30, uxtw]
    libdis_test+0x13c: ff 5b 7f f8  ldr xzr, [sp, wzr, uxtw #3]
    libdis_test+0x140: ff 7b 7f f8  ldr xzr, [sp, xzr, lsl #3]
    libdis_test+0x144: ff db 7f f8  ldr xzr, [sp, wzr, sxtw #3]
    libdis_test+0x148: ff fb 7f f8  ldr xzr, [sp, xzr, sxtx #3]
    libdis_test+0x14c: 00 68 60 3c  ldr b0, [x0, x0]
    libdis_test+0x150: ff 4b 7f 3c  ldr b31, [sp, wzr, uxtw]
    libdis_test+0x154: c0 5b 7e 3c  ldr b0, [x30, w30, uxtw #0]
    libdis_test+0x158: ff 7b 7f 3c  ldr b31, [sp, xzr, lsl #0]
    libdis_test+0x15c: c0 7b 7e 3c  ldr b0, [x30, x30, lsl #0]
    libdis_test+0x160: ff cb 7f 3c  ldr b31, [sp, wzr, sxtw]
    libdis_test+0x164: c0 db 7e 3c  ldr b0, [x30, w30, sxtw #0]
    libdis_test+0x168: ff eb 7f 3c  ldr b31, [sp, xzr, sxtx]
    libdis_test+0x16c: c0 fb 7e 3c  ldr b0, [x30, x30, sxtx #0]
    libdis_test+0x170: 00 68 60 7c  ldr h0, [x0, x0]
    libdis_test+0x174: ff 4b 7f 7c  ldr h31, [sp, wzr, uxtw]
    libdis_test+0x178: c0 5b 7e 7c  ldr h0, [x30, w30, uxtw #1]
    libdis_test+0x17c: ff 6b 7f 7c  ldr h31, [sp, xzr]
    libdis_test+0x180: c0 7b 7e 7c  ldr h0, [x30, x30, lsl #1]
    libdis_test+0x184: ff cb 7f 7c  ldr h31, [sp, wzr, sxtw]
    libdis_test+0x188: c0 db 7e 7c  ldr h0, [x30, w30, sxtw #1]
    libdis_test+0x18c: ff eb 7f 7c  ldr h31, [sp, xzr, sxtx]
    libdis_test+0x190: c0 fb 7e 7c  ldr h0, [x30, x30, sxtx #1]
    libdis_test+0x194: 00 68 60 bc  ldr s0, [x0, x0]
    libdis_test+0x198: ff 4b 7f bc  ldr s31, [sp, wzr, uxtw]
    libdis_test+0x19c: c0 5b 7e bc  ldr s0, [x30, w30, uxtw #2]
    libdis_test+0x1a0: ff 6b 7f bc  ldr s31, [sp, xzr]
    libdis_test+0x1a4: c0 7b 7e bc  ldr s0, [x30, x30, lsl #2]
    libdis_test+0x1a8: ff cb 7f bc  ldr s31, [sp, wzr, sxtw]
    libdis_test+0x1ac: c0 db 7e bc  ldr s0, [x30, w30, sxtw #2]
    libdis_test+0x1b0: ff eb 7f bc  ldr s31, [sp, xzr, sxtx]
    libdis_test+0x1b4: c0 fb 7e bc  ldr s0, [x30, x30, sxtx #2]
    libdis_test+0x1b8: 00 68 60 fc  ldr d0, [x0, x0]
    libdis_test+0x1bc: ff 4b 7f fc  ldr d31, [sp, wzr, uxtw]
    libdis_test+0x1c0: c0 5b 7e fc  ldr d0, [x30, w30, uxtw #3]
    libdis_test+0x1c4: ff 6b 7f fc  ldr d31, [sp, xzr]
    libdis_test+0x1c8: c0 7b 7e fc  ldr d0, [x30, x30, lsl #3]
    libdis_test+0x1cc: ff cb 7f fc  ldr d31, [sp, wzr, sxtw]
    libdis_test+0x1d0: c0 db 7e fc  ldr d0, [x30, w30, sxtw #3]
    libdis_test+0x1d4: ff eb 7f fc  ldr d31, [sp, xzr, sxtx]
    libdis_test+0x1d8: c0 fb 7e fc  ldr d0, [x30, x30, sxtx #3]
    libdis_test+0x1dc: 00 68 e0 3c  ldr q0, [x0, x0]
    libdis_test+0x1e0: ff 4b ff 3c  ldr q31, [sp, wzr, uxtw]
    libdis_test+0x1e4: c0 5b fe 3c  ldr q0, [x30, w30, uxtw #4]
    libdis_test+0x1e8: ff 6b ff 3c  ldr q31, [sp, xzr]
    libdis_test+0x1ec: c0 7b fe 3c  ldr q0, [x30, x30, lsl #4]
    libdis_test+0x1f0: ff cb ff 3c  ldr q31, [sp, wzr, sxtw]
    libdis_test+0x1f4: c0 db fe 3c  ldr q0, [x30, w30, sxtw #4]
    libdis_test+0x1f8: ff eb ff 3c  ldr q31, [sp, xzr, sxtx]
    libdis_test+0x1fc: c0 fb fe 3c  ldr q0, [x30, x30, sxtx #4]
    libdis_test+0x200: 00 68 20 78  strh w0, [x0, x0]
    libdis_test+0x204: de eb 3e 78  strh w30, [x30, x30, sxtx]
    libdis_test+0x208: ff fb 3f 78  strh wzr, [sp, xzr, sxtx #1]
    libdis_test+0x20c: ff 6b 3f 78  strh wzr, [sp, xzr]
    libdis_test+0x210: 00 68 60 78  ldrh w0, [x0, x0]
    libdis_test+0x214: de eb 7e 78  ldrh w30, [x30, x30, sxtx]
    libdis_test+0x218: ff fb 7f 78  ldrh wzr, [sp, xzr, sxtx #1]
    libdis_test+0x21c: ff 6b 7f 78  ldrh wzr, [sp, xzr]
    libdis_test+0x220: 00 68 e0 78  ldrsh w0, [x0, x0]
    libdis_test+0x224: de eb fe 78  ldrsh w30, [x30, x30, sxtx]
    libdis_test+0x228: ff fb ff 78  ldrsh wzr, [sp, xzr, sxtx #1]
    libdis_test+0x22c: ff 6b ff 78  ldrsh wzr, [sp, xzr]
    libdis_test+0x230: 00 68 a0 78  ldrsh x0, [x0, x0]
    libdis_test+0x234: de 4b be 78  ldrsh x30, [x30, w30, uxtw]
    libdis_test+0x238: ff 5b bf 78  ldrsh xzr, [sp, wzr, uxtw #1]
    libdis_test+0x23c: ff 7b bf 78  ldrsh xzr, [sp, xzr, lsl #1]
    libdis_test+0x240: ff db bf 78  ldrsh xzr, [sp, wzr, sxtw #1]
    libdis_test+0x244: ff fb bf 78  ldrsh xzr, [sp, xzr, sxtx #1]
    libdis_test+0x248: 00 68 a0 b8  ldrsw x0, [x0, x0]
    libdis_test+0x24c: de 4b be b8  ldrsw x30, [x30, w30, uxtw]
    libdis_test+0x250: ff 5b bf b8  ldrsw xzr, [sp, wzr, uxtw #2]
    libdis_test+0x254: ff 7b bf b8  ldrsw xzr, [sp, xzr, lsl #2]
    libdis_test+0x258: ff db bf b8  ldrsw xzr, [sp, wzr, sxtw #2]
    libdis_test+0x25c: ff fb bf b8  ldrsw xzr, [sp, xzr, sxtx #2]
